jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/lock/DebugRuntimeManagerLock.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.runtime.manager.impl.lock;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.jbpm.runtime.manager.spi.RuntimeManagerLock;
+
+
+public class DebugRuntimeManagerLock implements RuntimeManagerLock {
+
+    private ReentrantLock lock = new ReentrantLock(true);
+    private List<StackTraceElement[]> traces;
+    private long currentThreadId;
+
+    public DebugRuntimeManagerLock() {
+        traces = new ArrayList<>();
+        currentThreadId = -1;
+    }
+    
+    @Override
+    public void lock() {
+        lock.lock();
+        addTraces();
+    }
+
+    @Override
+    public boolean tryLock(long units, TimeUnit timeUnit) throws InterruptedException {
+        boolean outcome = lock.tryLock(units, timeUnit);
+        if(outcome) {
+            addTraces();
+        }
+        return outcome;
+    }
+
+    @Override
+    public void lockInterruptible() throws InterruptedException {
+        lock.lockInterruptibly();
+        addTraces();
+    }
+
+    private void addTraces() {
+        long newThreadId = Thread.currentThread().getId();
+        if(currentThreadId < 0 || currentThreadId != newThreadId) {
+            currentThreadId = newThreadId;
+            traces.clear();
+        }
+        // we add all stack traces during reentrant lock
+        traces.add(Thread.currentThread().getStackTrace());
```

Comment:
Can it even happen in the engine that we call lock() more than once in the same thread? By looking at the code when we first get the engine, it is locked and saved to the thread local (saveLocalRuntime method) and then it is reused within the same thread when calling getRuntimeEngine(), so we never reach the line with createLockOnGetEngine(). Just my observation.

Reply:
that is correct. Just trying to add the case of missuse of the runtime manager.